src/hotspot/share/code/codeCache.cpp

```diff
@@ -200,7 +200,10 @@ void CodeCache::initialize_heaps() {
   bool non_nmethod_set      = FLAG_IS_CMDLINE(NonNMethodCodeHeapSize);
   bool profiled_set         = FLAG_IS_CMDLINE(ProfiledCodeHeapSize);
   bool non_profiled_set     = FLAG_IS_CMDLINE(NonProfiledCodeHeapSize);
-  size_t min_size           = os::vm_page_size();
+  const size_t ps           = page_size(false, 8);
+  size_t min_size           = MAX3(os::vm_page_size(),
+                                   os::vm_allocation_granularity(),
+                                   ps);
```

Comment:
Is there a VM configuration in which `ps` could be smaller than `os::vm_page_size()`? If not, we could simplify this expression to `MAX2(os::vm_allocation_granularity(), ps)`.

Reply:
I am not sure, which is why I took the safe route and also included `os::vm_page_size()`. Maybe @jdksjolen knows if what Roberto suggests is a safe assumption?

Reply:
@tstuefe: @dafedafe recommended that we ask you for advice on the above (possibly related to https://github.com/openjdk/jdk/pull/14903/). Do you know the details?

Reply:
Without looking deep into this, something about this feels off. The original code using os::vm_page_size() should work fine, at least on non-Windows.

On almost all platforms all three values are identical. On AIX ps can be smaller than os::vm_page_size, and you probably don't care. On Windows, allocation granularity is 64k while pagesize is 4K.

Reply:
Thanks for the response Thomas! Indeed, the issue first surfaced on Windows where `os::vm_page_size()` < `os::vm_allocation_granularity()`. But, I have also been able to reproduce the issue on Linux with, e.g., `-XX:NMethodSizeLimit=351658240 -XX:+UseTransparentHugePages` which makes `ps` > `os::vm_page_size()`.

The proposed `MAX3` solution above is a safe approximation. However, it may also be safe to use `MAX2(os::vm_allocation_granularity(), ps)` (as Roberto suggests) if `os::vm_page_size()` is never the maximum. Even if `ps` >= `os::vm_page_size()` is not always true, maybe `os::vm_allocation_granularity()` >= `os::vm_page_size()` is always true?

Reply:
> maybe os::vm_allocation_granularity() >= os::vm_page_size() is always true?

I just checked and this holds on Windows (64 bits): `os::vm_page_size()` gives 4K and `os::vm_allocation_granularity()` gives 64K regardless of whether `UseLargePages` is set (this option affects the value of `ps` only). On all other OS:es we have `os::vm_allocation_granularity() == os::vm_page_size()` so it should be safe to remove `os::vm_page_size()`, right?
Perhaps you can remove it but add an assertion that it is always equal or smaller to `MAX2(os::vm_allocation_granularity(), ps)`.
Also, I think it would be good to include a Windows-specific test case to exercise the case where `os::vm_allocation_granularity() != os::vm_page_size()`, would this be possible?

Reply:
I'm fine with changing to MAX2 and adding an assert. There is already a test case that runs on all operating systems (including Windows where granularity != page size), that should be sufficient?